### ISSUE TITLE
Setting mission to fail delays the failure until next landing

### ISSUE THREAD

**Issue by Amazinite** [endless-sky/endless-sky#1938]
I believe this has been brought up before and it was intentional to stop a game crash, but maybe the mission should become invisible when it has been set to fail so that it actually looks as though it has failed, even if it still exists without your save file.

**Comment by KiLEdEnNis** [comment]
I think I saw 2 or 3 Wanderer missions that involved going to Hai space where you still see a "ghost" mission until you land

**Comment by Elyssaen** [comment]
_Oh_. I've been having real trouble with some invisible missions of mine that have a deadline of 1 and spawn an alien raid, which the Quarg come to reinforce you against. The most irritating problem has been that the Quarg follow you around for _ages_, despite `deadline 1`. I've got the issue on my to-do list for the stuff I'm working on but hadn't diagnosed it yet.

You've just cracked it for me. : ) But the fix I'd want would to actually be able to end the mission, because just making it invisible won't stop the odd following-about behaviour.

**Comment by Elyssaen** [comment]
Is it this commit, @Amazinite? 2a0b7b970191b23e5ac8580cc9fa5737dd2df1a5

@endless-sky, what do you recommend? If it's okay to half-diverge, for my situation too. (I was using `deadline 1` to try to get out of the Rim Archaeology 5A / Quarg Wardragon situation, where Gloraf Esa Kurayi can become a permanent companion as long as you don't land back at Zug. So my problem may have an alternative fix, one that'd work for Rim Archaeology [*], or it could be fixed if there was a safe way to end a failed mission before landing.)

As usual I'd be keen to contribute whatever seems like the best solution.

[*] Which would presumably be something like a "travelling to" characteristic for an NPC, which could be combined with `staying`... it's something neat I'd like to do anyway, but for me at least is a fair bit of work.

**Comment by Amazinite** [comment]
Yeah, that's what I recall. Although I'm not sure if before that commit, failing made missions disappear immediately. All I remember was trying to get missions to fail, it crashing, and that commit fixing it.

**Comment by NomadicVolcano** [comment]
Pull request #4764 should fix this bug planet-side: mission failure and completion status will be updated at landing and in the spaceport.  The game will not crash or hang as a result of this change.

This will *not* update mission status in space though, so your Quarg escorts will still follow you until you land.  Fixing that will require more extensive code changes.  You also need to consider what should happen to the Quarg escorts when the deadline runs out.

**Comment by Amazinite** [comment]
Done some further testing on this.

Say we have two missions, mission A and mission B. Here's what I've tested.

Situation | Outcome
---|---
Mission B is active. Mission A fails mission B on complete. | Mission B properly disappears.
Mission B is active. Mission A fails mission B on offer. | Mission B persists until next landing.
Mission B is active. Mission A fails mission B on accept. | Mission B persists until next landing.
Mission B is active. Mission A fails mission B on decline. | Mission B persists until next landing.
Mission A and B are active. Mission A fails mission B on abort. | Mission B persists until next landing.
Mission A fails itself on offer. | Mission A persists until the next landing.
Mission A fails itself on accept. | Mission A persists until the next landing.

If a mission is suppose to be failed but sticks around until next landing, its on fail action isn't run until the mission is gone, i.e. when you land.

In an instance where the mission persists until next landing, the mission will also be removed if you reload the pilot. What is happening is that the actions of a mission are adding the `failed` label to missions, but the mission doesn't check that `failed` label until PlayerInfo::StepMissions, which only occurs on landing. So what I can think to do would be to either make it so that after any action is done, call PlayerInfo::StepMissions (O(N^2) moment), or call PlayerInfo::StepMissions on PlayerInfo::TakeOff. The former seems to be the route that #4764 took (although only checking mission failure and completion after a mission was done offering, which I believe might have caused the cargo and passengers from said failed missions to persist until the next landing, even if the fail action was triggered), while the latter would be simpler but not allow lengthy chains of missions all failing each other for some reason.
